AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Notifications/AWSPinpointAnalyticsNotifications.swift

```diff
@@ -0,0 +1,356 @@
+            return metadata
+        }
+        
+        if let campaignMetadata = pinpointPayload[EventSource.campaign.rawValue] as? UserInfo {
+            metadata = (.campaign, campaignMetadata)
+            self.log.verbose("Found Pinpoint campaign with attributes: \(campaignMetadata)")
+        
+        } else if let journeyMetadata = pinpointPayload[EventSource.journey.rawValue] as? UserInfo {
+            metadata = (.journey, journeyMetadata)
+            self.log.verbose("Found Pinpoint journey with attributes: \(journeyMetadata)")
+        }
+        
+        if metadata.1 == nil {
+            fatalError("Pinpoint push notification payload not found.")
+        }
+        
+        return metadata
+    }
+    
+    private func addGlobalEventSourceMetadata(eventMetadata: UserInfo?,
+                                              eventSource: EventSource) async {
+        guard let eventMetadata = eventMetadata, !eventMetadata.isEmpty else {
+            return
+        }
+
+        // Remove previous global event source attributes from _globalAttributes
+        // only if event source type changes
+        // This is to prevent _globalAttributes containing attributes from multiple event sources (campaign/journey)
+        if eventSource != previousEventSource && eventSource != .unknown {
+            // remove all global attributes
+            await self.analyticsClient.removeAllGlobalEventSourceAttributes()
+            previousEventSource = eventSource
+        }
+        
+        do {
+            try await self.analyticsClient.setGlobalEventSourceAttributes(eventMetadata)
+        } catch {
+            self.log.error("Failed setting event source metadata")
+        }
+        
+        for (key, value) in eventMetadata {
+            guard let value = value as? String else {
+                log.debug("Skipping metadata with key \(key) because has a value of type \(type(of: value)).")
+                continue
+            }
+            await self.analyticsClient.addGlobalAttribute(value, forKey: key)
+        }
+        
+    }
+}
+
+// MARK: - AWSPinpointAnalyticsNotifications + handleDeepLink
+extension AWSPinpointAnalyticsNotifications {
+    typealias CanOpenURL = (URL) -> Bool
+    func handleDeepLinkForNotification(userInfo: UserInfo,
+                                       canOpenURL: CanOpenURL? = nil){
+#if canImport(UIKit)
+        let canOpenURL = canOpenURL ?? UIApplication.shared.canOpenURL
+#else
+        let canOpenURL = canOpenURL ?? {_ in false }
+#endif
+
+        guard let payload = pinpointPayloadFromNotificationPayload(notification: userInfo) as? [String: String],
+              let deepLink = payload[PinpointContext.Constants.Notifications.deeplinkKey],
+              let deepLinkURL = URL(string: deepLink) else {
+            return
+        }
+        if canOpenURL(deepLinkURL) {
+            DispatchQueue.main.async {
+                UIApplication.shared.open(deepLinkURL)
```

Comment:
`UIApplication` is a `UIKit` object

Reply:
oops I forgot to refactor that :)

Reply:
Updated to use `UIApplication.shared.open` only when UIKit is available, otherwise a noop closure is invoked